### PR TITLE
gherkin: Pass the content type of a docstring down into its pickle string form… 

### PR DESCRIPTION
**Pull request by rjwittams**
… for use in formatters.

## Summary

Pass the content type of a docstring down into its pickle string form. 
This allows formatters to use it. 
This used to work in Cucumber 1.2.5. 
A PR for cucumber-jvm accompanies this. 

## Motivation and Context

https://github.com/cucumber/cucumber-jvm/issues/1263
As suggested here by mpkorstanje , I am doing a PR against the mono repo.

## How Has This Been Tested?

Added in a unit test - not sure if it should be moved.
Also not sure that Gson-ing pickles is expected - given camelCase vs snake_case. 
Ran against my updated cucumber-jvm. Added a unit test there. 
Verified that my json report processor which works with 1.2.5 picks up the docstrings.

Ran all tests in cucumber-jvm. Examples: Pax-Exam : Calculator Test is broken before and after the change due to a missing deployment I think.

## Types of changes

- [x ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

Adds a field to PickleString. Not sure if that gets serialised within this project. 
Maintains the old constructor signature to avoid compile breaks. Could be deprecated.

## Checklist:

- [x] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

Not sure which if any docs should be updated - I think 1.2.5 preserved docs will now be correct.

### CODE REVIEW

**Comment by rjwittams** [comment]
https://github.com/cucumber/cucumber-jvm/pull/1265 <- this is the PR for cucumber-jvm

**Comment by rjwittams** [comment]
Updated the acceptance tests data. Used snake_case as I assumed that was the convention.

File: gherkin/java/src/main/java/gherkin/pickles/Compiler.java [hunk 1]
```diff
@@ -152,7 +152,8 @@ private void compileScenarioOutline(List<Pickle> pickles, List<PickleStep> backg
             result.add(
                     new PickleString(
                             pickleLocation(ds.getLocation()),
-                            interpolate(ds.getContent(), variableCells, valueCells)
+                            interpolate(ds.getContent(), variableCells, valueCells),
+                            ds.getContentType()
```
Comment (mpkorstanje) [inline gherkin/java/src/main/java/gherkin/pickles/Compiler.java]: Shouldn't this also be interpolated?

**Comment by rjwittams** [comment]
I hadn’t quite understood its significance. I see it’s for scenario outlines with examples. Can do so. What I’m not sure of if is the docstring content type can be used within a step. Eg can you use a PickleString as an argument in a step and get the whole thing rather than just its contents? It seems that it could be useful if not. Also may be more useful with the Transformers rework - could we discriminate transformers by content type.

Bit of a cross repo conversation ...

**Comment by mpkorstanje** [comment]
We're on slack btw. https://cucumberbdd-slack-invite.herokuapp.com/ the #commiters-jvm channel.

File: gherkin/java/testdata/good/docstrings.feature.pickles.ndjson [hunk 2]
```diff
@@ -1 +1 @@
...
...:[{"column":11,"line":4}],"text":"a simple DocString"},{"arguments":[{"content":"<foo>\n  <bar />\n</foo>","content_type":"xml","location":{"column":7,"line":12}}],"locations":[{"column":11,"line":11}],"text":"a DocString with content type"},{"arguments":[{"content":"wrongly indented line","location":{"column":7,"line":18}}],"locations":[{"column":9,"line":17}],"text":"a DocString with wrong indentation"},{"arguments":[{"content":"first line\nsecond line","location":{"column":7,"line":22}}],"locations":[{"column":9,"line":21}],"text":"a DocString with alternative separator"},{"arguments":[{"content":"first line\n\"\"\"\nthird line","location":{"column":7,"line":27}}],"locations":[{"column":9,"line":26}],"text":"a DocString with normal separator inside"},{"arguments":[{"content":"first line\n```\nthird line","location":{"column":7,"line":33}}],"locations":[{"column":9,"line":32}],"text":"a DocString with alternative separator inside"},{"arguments":[{"content":"first line\n\"\"\"\nthird line","location":{"column":7,"line":39}}],"locations":[{"column":9,"line":38}],"text":"a DocString with escaped separator inside"}],"tags":[]},"type":"pickle","uri":"testdata/good/docstrings.feature"}
```
Comment (brasmusson) [inline gherkin/java/testdata/good/docstrings.feature.pickles.ndjson]: This file should not be edited directly, it is [gherkin/testdata/good/docstrings.feature.pickles.ndjson](https://github.com/cucumber/cucumber/blob/master/gherkin/testdata/good/docstrings.feature.pickles.ndjson) which is the source file that is pushed down to each language version of the gherkin library (using [rsync_files](https://github.com/cucumber/cucumber/blob/master/scripts/functions.sh#L54-L63)). Since we use the same testdata for acceptance tests for all language version for the gherkin library, this pull-request cannot be merged until all 8 language versions have been updated.

**Comment by rjwittams** [comment]
Do you expect each contributor to fix all 8 languages, including those they’ve not used?

**Comment by brasmusson** [comment]
@rjwittams Many people can contribute to a pull-request, so "each contributor" does not need to fix all languages. However the acceptance test data should not diverge between language versions.

**Comment by brasmusson** [comment]
> And while the content type is not passed onto the step definition, ...

The Pickles are primarily for executing the test cases from each feature file, so if the content type is not needed for that, maybe the content type should not be added to the Pickles data.

A formatter can still access the content type, since the formatter has access to the complete feature file content.

**Comment by rjwittams** [comment]
Ie each reporter should reparse the feature file? Does that seem right?

What is the purpose of the content type? Given that it is not in the Pickle, it can not be used in the test execution. However, that is an arbitrary choice - argument transformers could make use of it. In fact I was planning to do this.

If there is no point in continuing let me know. It is however a clear regression in Cucumber JVM vs 1.2.5 .

**Comment by brasmusson** [comment]
> argument transformers could make use of it

This is a real argument for including the content type in the Pickle. That it is easier for the formatters is not. For instance the step keyword is not in the Pickle, even though practically every formatter needs to use it.

**Comment by rjwittams** [comment]
Also, as mentioned above, it may be useful (and consistent) for it to be transformed by interpolation.

File: gherkin/javascript/lib/gherkin/parser.js [hunk 3]
```diff
@@ -48,9 +48,6 @@ module.exports = function Parser(builder) {
     if(typeof tokenScanner == 'string') {
       tokenScanner = new TokenScanner(tokenScanner);
     }
-    if(typeof tokenMatcher == 'string') {
```
Comment (rjwittams) [inline gherkin/javascript/lib/gherkin/parser.js]: Not sure why this change was made in the generated parser - I assume it was missed out. I assume it is benign.

**Comment by rjwittams** [comment]
The build seemed to fail due to yarn timing out. Can it be restarted?
On binary and source compatibility: 
  C and go I did not try, as a struct size needs to change.
 Java and C# should have a compatible PickleString constructor defaulting content type to null.
 The rest are dynamic I think.

**Comment by mlvandijk** [comment]
@rjwittams Restarted the build. It fails with the following message:

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

**Comment by aslakhellesoy** [comment]
Hi @rjwittams,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber
